assignment2/2b/src/Tree.java

```diff
@@ -44,79 +50,71 @@ boolean add(T newValue, T parentVal, boolean isLeft) {
+        } else if (el2 == null) {
             throw new TreeException.ElementDoesntFound("element " + element2.toString() + " doesn't found");
         } else {
-            for (T i : parents1) {
-                if (parents2.contains(i)) {
-                    return i;
-                }
-            }
+            return getCommonAncestor(el1, el2);
         }
-
-        throw new TreeException.NoCommonAncestor("No common ancestor for values " + element1.toString()
-                + " and " + element2.toString());
     }
 
     /**
-     * Get list of parents for element. If element doesn't exist, return empty list
-     *
-     * @param element -- value of element
-     * @return -- list of ancestors
+     * Get common ancestor for two elements in the tree with known depth.
+     * @param el1 -- first element with depth.
+     * @param el2 -- second element with depth.
+     * @return -- value of common ancestor
      */
-    private List<T> getParents(T element) {
-        if (value.equals(element)) {
-            return getParents();
-        } else {
-            List<T> leftRes = left == null ? new LinkedList<T>() : left.getParents(element);
-            List<T> rightRes = right == null ? new LinkedList<T>() : right.getParents(element);
-            if (leftRes.isEmpty()) {
-                return rightRes;
-            } else {
-                return leftRes;
-            }
+    private T getCommonAncestor(Entry<Tree<T>, Integer> el1, Entry<Tree<T>, Integer> el2) {
+        while (el2.getValue() > el1.getValue()) {
+            el2 = new SimpleEntry<>(el2.getKey().parent, el2.getValue() - 1);
+        }
+        while (el1.getValue() > el2.getValue()) {
+            el1 = new SimpleEntry<>(el1.getKey().parent, el1.getValue() - 1);
+        }
+        while (!el1.getKey().value.equals(el2.getKey().value)) {
+            el1 = new SimpleEntry<>(el1.getKey().parent, el1.getValue() - 1);
+            el2 = new SimpleEntry<>(el2.getKey().parent, el2.getValue() - 1);
         }
+        return el1.getKey().value;
     }
 
     /**
-     * Get parents for current element
-     *
-     * @return list of ancestors
+     * Find element with depth
+     * @param element -- element to find
+     * @param depth -- current depth. 0 in the beginning
+     * @return -- element and depth
      */
-    private List<T> getParents() {
-        if (parent == null) {
-            return new LinkedList<T>();
-        } else {
-            List<T> res = parent.getParents();
-            res.add(0, parent.value);
-            return res;
+    private Entry<Tree<T>, Integer> findWithDepth(T element, int depth) {
+        if (element.equals(value)) {
+            return new SimpleEntry<>(this, depth);
         }
-    }
+        Entry<Tree<T>, Integer> rght = right == null ? null : right.findWithDepth(element, depth + 1);
+        Entry<Tree<T>, Integer> lft = left == null ? null : left.findWithDepth(element, depth + 1);
```

Comment:
This is not a good variable naming convention!

Even "left2" and "right2" would be better! Perhaps something like "rightResult" and "leftResult"?